<a href="https://colab.research.google.com/github/ytyimin/scm518/blob/main/Paint_Problem.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Paint Problem

**Objective and Prerequisites**

Gotham industries wants to maximize their profit by determining the number of gallons of paint they should produce. The objectives of the product-mix problem are:


*   Total usage of Base liquid must be less than total available.
*   Gallons of white base liquid produced must be at least twice the gallons of Blue base liquid
*   Ensure that the number of Gallons of paint produced are valid, i.e., non- negative production.



---
## **Problem Description**

![picture](https://drive.google.com/uc?export=view&id=1atih7hzkt4af3O4fA2FrOMU_arnP3yxy)

Gotham industries manufactures 2 types of paints white and blue. white paint requires 5 gallons of base liquid and blue paint requires 3 gallons of base liquid. White paint contributes 55 dollars profit and blue paint of 75 dollars profit. Marketing restrictions require that the gallons of white paint produced be at least twice the gallons of the blue paint produced. There are 3000 gallons of base liquid available.

---

## Model Formulation

### Indices:

$𝑖 \in \{W,B\}$ : Index to represent different paint types white and blue.

### Parameters: 

$p_i$, $i\in\{W,B\}$: unit profit of type $i$ paint (e.g., $p_W = \$55$, $p_B = \$75$)

$u_i$, $i\in\{W,B\}$: unsage of base liquid of type $i$ paint (e.g., $u_W = 2$ gallons, $u_B = \$5$ gallons)

$U$: available gallons of base liquid (e.g., $U = 3,000$ gallons) 

### Decisions Variables:

$x_i$, $i\in\{W,B\}$: gallons of type $i$ paint to produce

### Objective Function: 

\begin{equation}
\max_{x_i} \quad \sum_{i\in\{W,B\}} p_i*x_i
\tag{0}
\end{equation}   

### Constraints:

\begin{equation}
\sum_{i\in\{W,B\}} u_i*x_i \leq U \quad \text{(usage cannot exceed available)}
\tag{1}
\end{equation}   

\begin{equation}
x_W - 2*x_B \geq 0 \quad \text{(white paint more than twice of blue)}
\tag{2}
\end{equation}   

\begin{equation}
x_i \geq 0, i\in \{W,B\} \quad \text{(non-negative)}
\tag{3}
\end{equation}   

---

## Python Implementation

Import the Gurobi Python Module:

In [ ]:
# install Gurobi Python Module
%pip install gurobipy

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 12.9 MB 4.2 MB/s 


In [ ]:
# import necessary python libraries
from itertools import product
from math import sqrt, factorial
import numpy as np
import gurobipy as gp
from gurobipy import GRB

## Setup the model

In [ ]:
m = gp.Model() # Import and create the model

# Set the input Parameters: 
paint = [*range(0,2)]
paint_label =['W','B']

u = [5, 3]  # gallons of base liquid used to produce a particular paint colour

p = [55, 75]  # Profit for each paint colour sold 

U = 3000   # Total gallons of base liquid available for production

In [ ]:
# Set the decision variables
x = m.addVars(paint, vtype=GRB.CONTINUOUS, name='Produce')

In [ ]:
# Set the objective
m.setObjective(gp.quicksum(p[i]*x[i] for i in paint),  GRB.MAXIMIZE)

In [ ]:
# Set constraints

# Set total base liquid availability
ResourceConstrs = m.addConstr((gp.quicksum(u[i]*x[i] for i in paint) <= U), 
                                      name='ResourceConstrs')

# White paint two times blue paint
ProductionConstrs = m.addConstr(x[0] - 2* x[1] >= 0, 
                                      name='ProductionConstrs')

# Non-negative production
NonnegativeConstrs = m.addConstrs((x[i] >= 0 for i in paint), 
                                      name='ProductionConstrs')


## Solve the model:

In [ ]:
# Run the optimize solver
m.optimize()

Gurobi Optimizer version 10.0.0 build v10.0.0rc2 (linux64)

CPU model: Intel(R) Xeon(R) CPU @ 2.20GHz, instruction set [SSE2|AVX|AVX2]
Thread count: 1 physical cores, 2 logical processors, using up to 2 threads

Optimize a model with 4 rows, 2 columns and 6 nonzeros
Model fingerprint: 0x8a46a42a
Coefficient statistics:
  Matrix range     [1e+00, 5e+00]
  Objective range  [6e+01, 8e+01]
  Bounds range     [0e+00, 0e+00]
  RHS range        [3e+03, 3e+03]
Presolve removed 4 rows and 2 columns
Presolve time: 0.02s
Presolve: All rows and columns removed
Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    4.2692308e+04   0.000000e+00   0.000000e+00      0s

Solved in 0 iterations and 0.02 seconds (0.00 work units)
Optimal objective  4.269230769e+04


## Results of the Model:

In [ ]:
# print optimal production

print("\033[1m Optimal production and revenue by paint tyoes")
print("------------------------------------------\n")

total_production = 0
total_use = 0
# loop through all paints
for i in paint:
  print("\033[1m Paint",paint_label[i],"\033[0m : produce",round(x[i].x,2),"gallons, with a revenue of $", round(p[i]*x[i].x,2),"[base paint used:",round(u[i]*x[i].x,2),"]")
  total_production += x[i].x
  total_use += u[i]*x[i].x

print("------------------------------------------")
print("The total production is", round(total_production,2), "gallons, with a total revenue of",round(m.ObjVal,2),"[total base paint used:",round(total_use,2),"]")

 Optimal production and revenue by paint tyoes
------------------------------------------

 Paint W  : produce 461.54 gallons, with a revenue of $ 25384.62 [base paint used: 2307.69 ]
 Paint B  : produce 230.77 gallons, with a revenue of $ 17307.69 [base paint used: 692.31 ]
------------------------------------------
The total production is 692.31 gallons, with a total revenue of 42692.31 [total base paint used: 3000.0 ]


## **Conclusion**
In the above problem we determined the number of paints that need to be produced to obtain the maximum profit keeping in mind the Total usage of Base liquid must be less than total available.Gallons of white base liquid produced must be at least twice the gallons of Blue base liquid constarints. It is seen that Gotham industries should manufacture 462 gallons of white paint and 231 gallons of blue paint to maximize it's profit which comes out to be $42,692

## **References**
[1] Sixty examples of business optimization models. https://ytyimin.github.io/tart-cherry/.

[2] Gurobi python reference. https://www.gurobi.com/documentation/

[3] This example is developed by

MD shahnaz ahamadi  www.linkedin.com/in/shahnazahamadi

Manohar K www.linkedin.com/in/manohar-k/

Hemanth Gopinath www.linkedin.com/in/hemanthgopinath/